README.md
# Terraformize

Apply\Destory Terraform modules via a simple REST API endpoint.

Github actions CI unit tests & auto dockerhub push status: [![CI/CD](https://github.com/Perteghella/terraformize/actions/workflows/full_ci_cd_workflow.yml/badge.svg)](https://github.com/Perteghella/terraformize/actions/workflows/full_ci_cd_workflow.yml)

Code coverage: [![codecov](https://codecov.io/gh/Perteghella/terraformize/branch/master/graph/badge.svg)](https://codecov.io/gh/Perteghella/terraformize)

# Features

* REST API to run:
    * `terraform apply`
    * `terraform destroy`
    * `terraform plan`
    * `terraform output`
* No code changes needed, supports 100% of all terraform modules unmodified
* Built in support for multiple terraform workspaces
* Can pass variables to the terraform run via the request body (passed as a -var arg to the `terraform apply` or `terraform destroy` command)
* Supports multiple module directories
* Automatically runs terraform init before changes
* Returned response includes all the logs of stdout & stderr of terraform for easy debugging
* Stateless (requires you use a non local terraform backend)
* Containerized
* Health check endpoint included
* support all terraform backends that support multiple workspaces
* No DB needed, all data stored at the terraform backend of your choosing
* terraformize scales out as much as you need risk-free (requires you use a backend that support state locking)
* AMD64, Arm & Arm64 support
* (Optional) Ability to have Terraform run on a separate thread and have it return the terraform run result to a (configurable per request) webhook in a non-blocking way
* (Optional) Ability to work off a RabbitMQ queue and have it return the terraform run result to a different queue

# Possible use cases

* Setting up SaaS clusters\products\etc for clients in a fully automatic way
    * Each customer gets his own workspace and you just run the needed modules to create that customer products
    * Devs don't have to know how the terraform module works, they just point to a rest API & know it creates everything they need
* CI/CD integration
    * You can consider it as a terraform worker that is very easy to turn on via your CI/CD tool as it's just a REST request away
* Automatic system creation and\or scaling
    * Easy to intergrate to autoscalers as it can run the same module with diffrent variables passed to the `terraform apply` command via the request body to scale services up\down as needed
    * Easy to give your company employees a self-service endpoint and have them create and\or remove infrastracutre themselves when it's just an API request away
* Seperate who writes the terraform modules to who use them easily

# Running

Running Terraformize is as simple as running a docker container

```docker
docker run -d -p 80:80 -v /path/to/my/terraform/module/dir:/www/terraform_modules/ perteghella/terraformize
```

Feel free to skip to the end of the document for a working example that will explain how to use Terraformize

# Configuration options

Terraformize uses sane defaults but they can all be easily changed:

| value                          | envvar                        | default value            | notes                                                                                                                   |
|--------------------------------|-------------------------------|--------------------------|-------------------------------------------------------------------------------------------------------------------------|
| basic_auth_user                | BASIC_AUTH_USER               | None                     | Basic auth username to use                                                                                              |
| basic_auth_password            | BASIC_AUTH_PASSWORD           | None                     | Basic auth password to use                                                                                              |
| auth_token                     | AUTH_TOKEN                    | None                     | bearer token to use                                                                                                     |
| terraform_binary_path          | TERRAFORM_BINARY_PATH         | None                     | The path to the terraform binary, if None will use the default OS PATH to find it                                       |
| terraform_modules_path         | TERRAFORM_MODULES_PATH        | /www/terraform_modules   | The path to the parent directory where all terraform module directories will be stored at as subdirs                    |
| parallelism                    | PARALLELISM                   | 10                       | The number of parallel resource operations                                                                              |
| rabbit_url_connection_string   | RABBIT_URL_CONNECTION_STRING  | None                     | The URL paramters string to connect to RabbitMQ with, if unset RabbitMQ will not be used and only API will be possible  |
| rabbit_read_queue              | RABBIT_READ_QUEUE             | terraformize_read_queue  | Name of the queue to read messages from                                                                                 |
| rabbit_reply_queue             | RABBIT_REPLY_QUEUE            | terraformize_reply_queue | Name of the queue to respond with the run result to                                                                     |
|                                | CONFIG_DIR                    | /www/config              | The path to the directory where configuration files are stored at                                                       |
|                                | HOST                          | 0.0.0.0                  | The IP for gunicorn to bind to                                                                                          |
|                                | PORT                          | 80                       | The port for gunicorn to bind to                                                                                        |
|                                | WORKER_CLASS                  | sync                     | The gunicorn class to use                                                                                               |
|                                | WORKERS                       | 1                        | Number of gunicorn workers                                                                                              |
|                                | THREADS                       | 1                        | Number of gunicorn threads                                                                                              |
|                                | PRELOAD                       | False                    | If gunicorn should preload the code                                                                                     |
|                                | LOG_LEVEL                     | error                    | The log level for gunicorn                                                                                              |
|                                | TIMEOUT                       | 600                      | The timeout for gunicorn, if your terraform run takes longer you will need to increase it                               |


The easiest way to change a default value is to pass the envvar key\value to the docker container with the `-e` cli arg but if you want you can also create a configuration file with the settings you wish (in whatever of the standard format you desire) & place it in the /www/config folder inside the container.

Most providers also allow setting their configuration access_keys\etc via envvars use `-e` cli args to configure them is ideal as well but should you wish to configure a file you can also easily mount\copy it into the container as well.

## Authentication 

Terraformize supports 3 authentication methods:

* Basic auth - will require you to pass a `Authorization Basic your_user_pass_base64_combo` header with your_user_pass_base64_combo being the same as `basic_auth_user` & `basic_auth_password` configured in Terraformize 
* Bearer auth - will require you to pass a `Authorization Bearer your_token` header with your_token being the same as the `auth_token` configured in Terraformize
* No auth - will be used if both the Basic auth & Bearer auth are disabled, note that the /v1/health health-check point never requires authentication

# Endpoints

* POST /v1/module_folder_name/workspace_name
    * runs `terraform apply` for you
    * takes care of auto approval of the run, auto init & workspace switching as needed
    * takes variables which are passed to `terraform apply` as a JSON in the body of the message in the format of `{"var_key1": "var_value1", "var_key2": "var_value2"}`
    * Returns 200 HTTP status code if everything is ok, 404 if you gave it a non existing module_folder_name path & 400 if the `terraform apply` ran but failed to make all needed modifications
    * Also returns a JSON body of `{"init_stdout": "...", "init_stderr": "...", "stderr": "...", "stdout": "..."}` with the stderr & stdout of the `terraform apply` & `terraform init` run
    * If you pass a `webhook` URL paramter  with the address of the webhook terraformize will return a `202` HTTP code with a body of `{{'request_uuid': 'ec743bc4-0724-4f44-9ad3-5814071faddx'}}` to the request then work behind the scene to run terraform in a non blocking way, to result of the terraform run will be sent to the webhook address you configured along with the UUID of the request for you to know which request said result related to
* DELETE /v1/module_folder_name/workspace_name
    * runs `terraform destroy` for you
    * takes care of auto approval of the run, auto init & workspace switching as needed
    * takes variables which are passed to `terraform destroy` as a JSON in the body of the message in the format of `{"var_key1": "var_value1", "var_key2": "var_value2"}`
    * Returns 200 HTTP status code if everything is ok, 404 if you gave it a non existing module_folder_name path & 400 if the `terraform destroy` ran but failed to make all needed modifications
    * Also returns a JSON body of `{"init_stdout": "...", "init_stderr": "...", "stderr": "...", "stdout": "..."}` with the stderr & stdout of the `terraform destroy` & `terraform init` run
    * In order to preserve the history of terraform runs in your backend the workspace is not deleted automatically, only the infrastructure is destroyed
    * If you pass a `webhook` URL paramter  with the address of the webhook terraformize will return a `202` HTTP code with a body of `{{'request_uuid': 'ec743bc4-0724-4f44-9ad3-5814071faddx'}}` to the request then work behind the scene to run terraform in a non blocking way, to result of the terraform run will be sent to the webhook address you configured along with the UUID of the request for you to know which request said result related to
* POST /v1/module_folder_name/workspace_name/plan
    * runs `terraform plan` for you
    * takes care of auto approval of the run, auto init & workspace switching as needed
    * takes variables which are passed to `terraform apply` as a JSON in the body of the message in the format of `{"var_key1": "var_value1", "var_key2": "var_value2"}`
    * Returns 200 HTTP status code if everything is ok, 404 if you gave it a non existing module_folder_name path & 400 if the `terraform apply` ran but failed to plan all needed modifications
    * Also returns a JSON body of `{"init_stdout": "...", "init_stderr": "...", "stderr": "...", "stdout": "...", "exit_code": "0""}` with the stderr & stdout of the `terraform apply` & `terraform init` run
    * If you pass a `webhook` URL paramter  with the address of the webhook terraformize will return a `202` HTTP code with a body of `{{'request_uuid': 'ec743bc4-0724-4f44-9ad3-5814071faddx'}}` to the request then work behind the scene to run terraform in a non blocking way, to result of the terraform run will be sent to the webhook address you configured along with the UUID of the request for you to know which request said result related to
* GET /v1/health
    * Returns 200 HTTP status code
    * Also returns a JSON body of {"healthy": true}
    * Never needs auth
    * Useful to monitoring the health of Terraformize service

# RabbitMQ queue

if you prefer using RabbitMQ instead of the API then you'll need to configure the `rabbit_url_connection_string` (examples can be seen at https://pika.readthedocs.io/en/stable/examples/using_urlparameters.html#using-urlparameters), Terrafromize will then use 2 Queues on rabbit (defined at the `rabbit_read_queue` & `rabbit_reply_queue` params), you don't have to create the queues manually, if need be they will be created.

Now all you need to do in order to have a terraform run is to publish a message to the `rabbit_read_queue` with the following format:

```json
{
  "module_folder": "module_folder_name",
  "workspace": "workspace_name",
  "uuid": "unique_uuid_you_created_to_identify_the_request",
  "run_type": "apply/destroy/plan",
  "run_variables": {
    "var_to_pass_to_terraform_key": "var_to_pass_to_terraform_value",
    "another_var_to_pass_to_terraform_key": "another_var_to_pass_to_terraform_value"
  }
}
```

Terraformize will then run terraform for you and will return the result of the terraform run to the `rabbit_reply_queue` queue in the following format:

```json
{
  "uuid": "unique_uuid_you_created_to_identify_the_request",
  "init_stdout": "...", 
  "init_stderr": "...", 
  "stderr": "...", 
  "stdout": "...", 
  "exit_code": 0
}
```

It's up to you to ensure the `uuid` you pass is indeed unique.

# Example

1. First we will need a terraform module so create a folder named terraformize_test:
    ```shell script
    mkdir terraformize_test
    ```
   Make sure **not** to `cd` into the folder as we will be mounting it into the container from the parent folder in a couple of steps
2. Now we need a valid terraform configuration in it, if it works in terraform it will work with terraformize but for this example we will keep it simple with a single `terraformize_test/test.tf` file:
    ```hcl-terraform
    resource "null_resource" "test" {
      count   = 1
    }
    
    variable "test_var" {
      description = "an example variable"
      default = "my_variable_default_value"
    }
    
    output "test" {
      value = var.test_var
    }
    ```
3. We will also need to add the folder we created into the Terraformize container, this can be done by many different way (for example creating a container that copies our modules into a new image with the FROM base image being Terraformize base image) but for this example we will simply mount the folder path into the container as we run it:
    ```docker
    docker run -d -p 80:80 -v `pwd`:/www/terraform_modules naorlivne/terraformize
    ```
4. Now we can run the terraform module by simply calling it which will run `terraform apply` for us (notice how we are passing variables in the body):
    ```shell script
    curl -X POST \
      http://127.0.0.1/v1/terraformize_test/my_workspace \
      -H 'Content-Type: application/json' \
      -H 'cache-control: no-cache' \
      -d '{
        "test_var": "hello-world"
    }'
    ```
5. And lets create another copy infra of the same module in another workspace:
    ```shell script
    curl -X POST \
      http://127.0.0.1/v1/terraformize_test/my_other_workspace \
      -H 'Content-Type: application/json' \
      -H 'cache-control: no-cache' \
      -d '{
        "test_var": "hello-world"
    }'
    ```
6. Now that we are done let's delete them both (this will run `terrafrom destroy` for us):
    ```shell script
   curl -X DELETE \
      http://127.0.0.1/v1/terraformize_test/my_workspace \
      -H 'Content-Type: application/json' \
      -H 'cache-control: no-cache' \
      -d '{
        "test_var": "hello-world"
    }' 
   curl -X DELETE \
      http://127.0.0.1/v1/terraformize_test/my_other_workspace \
      -H 'Content-Type: application/json' \
      -H 'cache-control: no-cache' \
      -d '{
        "test_var": "hello-world"
    }'
    ```


# Build the docker image manually

Build and push to docker hub a multi arch image

```
docker buildx build  --platform linux/amd64,linux/arm64 -f Dockerfile -t perteghella/terraformize:1.1.0 -t perteghella/terraformize:latest . --push
docker buildx build  --platform linux/amd64,linux/arm64 -f Dockerfile -t registry.gitlab.com/perteghella/terraformize:1.1.0 -t registry.gitlab.com/perteghella/terraformize:latest . --push
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.1.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

Types of changes

- Added - for new features.
- Changed - for changes in existing functionality.
- Deprecated - for soon-to-be removed features.
- Removed - for now removed features.
- Fixed - for any bug fixes.
- Security - in case of vulnerabilities.

## [1.1.0] - 2023-04-21

### Changed

- Single Dockerfile for multi-architecture
- Update to terraform 1.4.5

## [1.0.0] - 2023-03-19

### Changed

- Forked the project https://github.com/naorlivne/terraformize
- update to terraform 1.4.2 and python:3.11.2-alpine3.17
- dockerignore and gitignore to skip mac/osx files

### Added

- CHANGELOG.md this file

terraformize/terraformize_endpoint.py
from terraformize.terraformize_terraform_wrapper import *
from terraformize.terraformize_configure import *
from terraformize.terraformize_webhook import *
from flask import Flask, request, jsonify
from flask_httpauth import HTTPBasicAuth, HTTPTokenAuth, MultiAuth
from typing import Optional, Tuple
from threading import Thread


API_VERSION = "v1"


app = Flask(__name__)
basic_auth = HTTPBasicAuth(realm='terraformize')
token_auth = HTTPTokenAuth('Bearer')
multi_auth = MultiAuth(basic_auth, token_auth)


def long_running_task(**kwargs):
    """
    Used in threading option to run the terraform apply/plan/destroy commands on in it's own thread and return the
    response to a webhook

    Arguments:
        :param command: plan/destroy/apply - the terraform command to run
        :param variables: the body of the variables to pass to the terraform command
        :param module_path: the path to the module to run
        :param workspace_name: name of terraform workspace
        :param terraform_request_uuid: the uuid of the long running request
        :param webhook_url: url to return the results of the terraform run once it completes
    """
    command = kwargs.get('command', {})
    variables = kwargs.get('variables', {})
    module_path = kwargs.get('module_path', {})
    workspace_name = kwargs.get('workspace_name', {})
    terraform_request_uuid = kwargs.get('terraform_request_uuid', {})
    webhook_url = kwargs.get('webhook_url', {})
    terraform_object = Terraformize(workspace_name, configuration["terraform_modules_path"] + "/" + module_path,
                                    terraform_bin_path=configuration["terraform_binary_path"])
    if command == "plan":
        terraform_return_code, terraform_stdout, terraform_stderr = terraform_object.plan(
            variables, configuration["parallelism"]
        )
    elif command == "apply":
        terraform_return_code, terraform_stdout, terraform_stderr = terraform_object.apply(
            variables, configuration["parallelism"]
        )
    elif command == "destroy":
        terraform_return_code, terraform_stdout, terraform_stderr = terraform_object.destroy(
            variables, configuration["parallelism"]
        )
    send_webhook_result(webhook_url, terraform_object.init_stdout, terraform_object.init_stderr, terraform_stdout,
                        terraform_stderr, terraform_return_code, terraform_request_uuid)


def terraform_return_code_to_http_code(terraform_return_code: int, plan_mode=False) -> int:
    """
    Converts the exit code given by terraform to HTTP return code

    Arguments:
        :param terraform_return_code: the exit code given by terraform
        :param plan_mode: default to False, if True will assume the result is exit code from "terraform plan" which is
        a bit different the standard TF exit codes

    Returns:
        :return return_code: 200 if terraform exit code was 0 (or 2 in plan mode), 400 otherwise
    """
    if terraform_return_code == 0 or (plan_mode is True and terraform_return_code == 2):
        return_code = 200
    else:
        return_code = 400
    return return_code


@basic_auth.verify_password
def verify_password(username: Optional[str], password: Optional[str]) -> bool:
    """
    Authenticate the basic_auth user & password given by the end-user

    Arguments:
        :param username: the username to check
        :param password: the password to check

    Returns:
        :return: True if the user&pass is correct or auth is disabled, False otherwise
    """
    # if auth_enabled is set to false then always allow access
    if configuration["auth_enabled"] is False:
        return True
    # else if username and password matches the admin user set in the manager config allow access
    elif username == configuration["basic_auth_user"] and password == configuration["basic_auth_password"]:
        return True
    # in any other case return false
    else:
        return False


# this function checks token based auth to allow access to authenticated users.
@token_auth.verify_token
def verify_token(token: Optional[str]) -> bool:
    """
    Authenticate the bearer token given by the end-user

    Arguments:
        :param token: the token to check

    Returns:
        :return: True if the token is correct or auth is disabled, False otherwise
    """
    # if auth_enabled is set to false then always allow access
    if configuration["auth_enabled"] is False:
        return True
    # else if the token matches the admin user set in the manager config allow access
    elif configuration["auth_token"] == token:
        return True
    # in any other case return false
    else:
        return False


@app.route('/' + API_VERSION + '/<module_path>/<workspace_name>/plan', methods=["POST"])
@multi_auth.login_required
def plan_terraform(module_path: str, workspace_name: str) -> Tuple[str, int]:
    """
    A REST endpoint to plan terraform modules at a given module path inside the main module directory path at a given
    workspace, if the "webhook" is set to a URL the function will work in a threaded format and return only the request
    UUID then will return the full request response once that completes in a non blocking format

    Arguments:
        :param module_path:  the name of the subdirectory for the module inside the "terraform_modules_path" to run
        "terraform plan" at
        :param workspace_name: the name of the workspace to run "terraform plan" at

    Returns:
        :return return_body: a JSON of the terraform exit code, stdout & stderr from the terraform run, if a arg of
        "webhook" is  passed will only return the UUID of the request
        :return terraform_return_code: the terraform return code

    Exceptions:
        :except FileNotFoundError: will return HTTP 404 with a JSON of the stderr it catch from "terraform init" or
        "terraform plan"
    """
    try:
        webhook = request.args.get('webhook', None)
        if webhook is None:
            terraform_object = Terraformize(workspace_name, configuration["terraform_modules_path"] + "/" + module_path,
                                            terraform_bin_path=configuration["terraform_binary_path"])
            terraform_return_code, terraform_stdout, terraform_stderr = terraform_object.plan(
                request.get_json(silent=True), configuration["parallelism"]
            )
            return_body = jsonify({
                "init_stdout": terraform_object.init_stdout,
                "init_stderr": terraform_object.init_stderr,
                "stdout": terraform_stdout,
                "stderr": terraform_stderr,
                "exit_code": terraform_return_code
            })
            terraform_return_code = terraform_return_code_to_http_code(int(terraform_return_code), plan_mode=True)
            return return_body, terraform_return_code
        else:
            terraform_request_uuid, terraform_request_uuid_json = create_request_uuid()
            thread = Thread(target=long_running_task, kwargs={'command': "plan",
                                                              'variables': request.get_json(silent=True),
                                                              'workspace_name': workspace_name,
                                                              'module_path': module_path,
                                                              'terraform_request_uuid': terraform_request_uuid,
                                                              'webhook_url': webhook})
            thread.start()
            return terraform_request_uuid_json, 202

    except FileNotFoundError as error_log:
        return jsonify({"error": str(error_log)}), 404


@app.route('/' + API_VERSION + '/<module_path>/<workspace_name>/output', methods=["POST"])
@multi_auth.login_required
def output_terraform(module_path: str, workspace_name: str) -> Tuple[str, int]:
    """
    A REST endpoint to OUTPUT terraform modules at a given module path inside the main module directory path at a given
    workspace, if the "webhook" is set to a URL the function will work in a threaded format and return only the request
    UUID then will return the full request response once that completes in a non blocking format

    Arguments:
        :param module_path:  the name of the subdirectory for the module inside the "terraform_modules_path" to run
        "terraform plan" at
        :param workspace_name: the name of the workspace to run "terraform plan" at

    Returns:
        :return return_body: a JSON of the terraform exit code, stdout & stderr from the terraform run, if a arg of
        "webhook" is  passed will only return the UUID of the request
        :return terraform_return_code: the terraform return code

    Exceptions:
        :except FileNotFoundError: will return HTTP 404 with a JSON of the stderr it catch from "terraform init" or
        "terraform plan"
    """
    try:
        webhook = request.args.get('webhook', None)
        if webhook is None:
            terraform_object = Terraformize(workspace_name, configuration["terraform_modules_path"] + "/" + module_path,
                                            terraform_bin_path=configuration["terraform_binary_path"])
            terraform_return_code, terraform_stdout, terraform_stderr = terraform_object.plan(
                request.get_json(silent=True), configuration["parallelism"]
            )
            return_body = jsonify({
#                "init_stdout": terraform_object.init_stdout,
#                "init_stderr": terraform_object.init_stderr,
                "stdout": terraform_stdout,
#                "stderr": terraform_stderr,
                "exit_code": terraform_return_code
            })
            terraform_return_code = terraform_return_code_to_http_code(int(terraform_return_code), plan_mode=True)
            return return_body, terraform_return_code
        else:
            terraform_request_uuid, terraform_request_uuid_json = create_request_uuid()
            thread = Thread(target=long_running_task, kwargs={'command': "output",
                                                              'variables': request.get_json(silent=True),
                                                              'workspace_name': workspace_name,
                                                              'module_path': module_path,
                                                              'terraform_request_uuid': terraform_request_uuid,
                                                              'webhook_url': webhook})
            thread.start()
            return terraform_request_uuid_json, 202

    except FileNotFoundError as error_log:
        return jsonify({"error": str(error_log)}), 404



@app.route('/' + API_VERSION + '/<module_path>/<workspace_name>', methods=["POST"])
@multi_auth.login_required
def apply_terraform(module_path: str, workspace_name: str) -> Tuple[str, int]:
    """
    A REST endpoint to apply terraform modules at a given module path inside the main module directory path at a given
    workspace, if the "webhook" is set to a URL the function will work in a threaded format and return only the request
    UUID then will return the full request response once that completes in a non blocking format

    Arguments:
        :param module_path:  the name of the subdirectory for the module inside the "terraform_modules_path" to run
        "terraform apply" at
        :param workspace_name: the name of the workspace to run "terraform apply" at

    Returns:
        :return return_body: a JSON of the stdout & stderr from the terraform run
        :return terraform_return_code: the terraform return code

    Exceptions:
        :except FileNotFoundError: will return HTTP 404 with a JSON of the stderr it catch from "terraform init" or
        "terraform apply"
    """
    try:
        webhook = request.args.get('webhook', None)
        if webhook is None:
            terraform_object = Terraformize(workspace_name, configuration["terraform_modules_path"] + "/" + module_path,
                                            terraform_bin_path=configuration["terraform_binary_path"])
            terraform_return_code, terraform_stdout, terraform_stderr = terraform_object.apply(
                request.get_json(silent=True), configuration["parallelism"]
            )
            return_body = jsonify({
                "init_stdout": terraform_object.init_stdout,
                "init_stderr": terraform_object.init_stderr,
                "stdout": terraform_stdout,
                "stderr": terraform_stderr
            })
            terraform_return_code = terraform_return_code_to_http_code(int(terraform_return_code))
            return return_body, terraform_return_code
        else:
            terraform_request_uuid, terraform_request_uuid_json = create_request_uuid()
            thread = Thread(target=long_running_task, kwargs={'command': "apply",
                                                              'variables': request.get_json(silent=True),
                                                              'workspace_name': workspace_name,
                                                              'module_path': module_path,
                                                              'terraform_request_uuid': terraform_request_uuid,
                                                              'webhook_url': webhook})
            thread.start()
            return terraform_request_uuid_json, 202
    except FileNotFoundError as error_log:
        return jsonify({"error": str(error_log)}), 404


@app.route('/' + API_VERSION + '/<module_path>/<workspace_name>', methods=["DELETE"])
@multi_auth.login_required
def destroy_terraform(module_path: str, workspace_name: str) -> Tuple[str, int]:
    """
    A REST endpoint to destroy terraform modules at a given module path inside the main module directory path at a given
    workspace, if the "webhook" is set to a URL the function will work in a threaded format and return only the request
    UUID then will return the full request response once that completes in a non blocking format

    Arguments:
        :param module_path:  the name of the subdirectory for the module inside the "terraform_modules_path" to run
        "terraform destroy" at
        :param workspace_name: the name of the workspace to run "terraform destroy" at

    Returns:
        :return return_body: a JSON of the stdout & stderr from the terraform run
        :return terraform_return_code: the terraform return code

    Exceptions:
        :except FileNotFoundError: will return HTTP 404 with a JSON of the stderr it catch from "terraform init" or
        "terraform destroy"
    """
    try:
        webhook = request.args.get('webhook', None)
        if webhook is None:
            terraform_object = Terraformize(workspace_name, configuration["terraform_modules_path"] + "/" + module_path,
                                            terraform_bin_path=configuration["terraform_binary_path"])
            terraform_return_code, terraform_stdout, terraform_stderr = terraform_object.destroy(
                request.get_json(silent=True), configuration["parallelism"]
            )
            return_body = jsonify({
                "init_stdout": terraform_object.init_stdout,
                "init_stderr": terraform_object.init_stderr,
                "stdout": terraform_stdout,
                "stderr": terraform_stderr
            })
            terraform_return_code = terraform_return_code_to_http_code(int(terraform_return_code))
            return return_body, terraform_return_code
        else:
            terraform_request_uuid, terraform_request_uuid_json = create_request_uuid()
            thread = Thread(target=long_running_task, kwargs={'command': "destroy",
                                                              'variables': request.get_json(silent=True),
                                                              'workspace_name': workspace_name,
                                                              'module_path': module_path,
                                                              'terraform_request_uuid': terraform_request_uuid,
                                                              'webhook_url': webhook})
            thread.start()
            return terraform_request_uuid_json, 202
    except FileNotFoundError as error_log:
        return jsonify({"error": str(error_log)}), 404


@app.route('/' + API_VERSION + '/health', methods=["GET"])
def health_check() -> Tuple[str, int]:
    """
    A REST endpoint to make sure that terraformize is working

    Returns:
        :return return_body: a JSON of {"healthy": true}
        :return terraform_return_code: 200

    """
    return jsonify({"healthy": True}), 200
